### ISSUE TITLE
删除自定义策略失效的问题（On the Invalidation of Delete Custom Policy）

### ISSUE THREAD

**Issue by helloworld0929** [marcosbarbero/spring-cloud-zuul-ratelimit#199]
你好，在测试的过程中遇见了一个问题，但我不知道问题出在哪里。
问题描述：在yml的配置里面设置了一个自定义的限流策略，部分代码如下：
（Hello, I met a problem during the test, but I don't know what the problem is.
Description of the problem: In the configuration of yml, a custom current limiting policy is set up. Some codes are as follows:）
```yaml
zuul:
  routes:
    testservice:
      path: /test/**
  ratelimit:
    enabled: true
    repository: redis
    policy-list:
      testservice:
        - limit: 2 
          quota: 1
          refresh-interval: 3
```
在测试的过程中时生效的，但是当我将下面的代码修改为如下，尝试测试删除自定义策略时，出现了自定义的限流策略仍然生效的问题。
（It worked during the test, but when I changed the following code to try to test the Delete Custom Policy, there was a problem that the Custom Current Limitation Policy still worked.）
```yaml
zuul:
#   routes:
#   testservice:
#      path: /test/**
  ratelimit:
    enabled: true
    repository: redis
    policy-list:
#       testservice:
#         - limit: 2 
#           quota: 1
#           refresh-interval: 3
```
也就是说，我这里删除了自定义的限流策略，再次尝试测试时，发现，上一次的自定义限流策略仍然有效...什么鬼？
（That is to say, I deleted the custom current limiting strategy here, and when I tried to test again, I found that the last custom current limiting strategy was still valid... What the hell?）
就这些，谢谢。
（Thats all，thank you.）

**Comment by marcosbarbero** [comment]
Hi, here some questions for you:

- How did you test it? 
- Where is this file placed, is it an external or an internal file?
- Did you restart the application?

Have in mind that all the configuration relies on the spring-boot default properties load, meaning that if something went wrong, the problem is improbable to be on this project.

**Comment by helloworld0929** [comment]
你好，我的回答如下：（Hello, my answer is as follows:）

1. 使用linux下的命令行，curl http...测试（Using the command line under linux, curl http... test）
2. application.yml 在spring cloud config 的配置中心（application.yml file in spring cloud config center）
3. 请看第二个回答（as answer 2）
哈哈哈，你能看出我的问题出在哪里了吗？
（Haha, can you find out where my problem is?）

**Comment by marcosbarbero** [comment]
I don't understand your answer to the third question, did you restart your application? As you are using Spring Cloud Config, did you refresh the context?

Unfortunately, I can't tell where's your problem but I'm pretty sure that using Spring Cloud Config with this project wouldn't be a problem as it supports [@RefreshScope](https://github.com/marcosbarbero/spring-cloud-zuul-ratelimit/blob/master/spring-cloud-zuul-ratelimit-core/src/main/java/com/marcosbarbero/cloud/autoconfigure/zuul/ratelimit/config/properties/RateLimitProperties.java#L45)

**Comment by helloworld0929** [comment]
使用了Redis作为限流数据存储库，很感谢您解决了refresh-interval无法刷新的问题，我现在遇见的问题是，当我自定义一个限流策略，测试可行，马上修改配置文件删除当前的自定义限流配置，并推送至spring cloud config配置中心，发现已经删除的自定义的限流策略任然可用，啊偶~
（Redis is used as the current limiting data repository. Thank you for resolving the problem that refresh-interval cannot be refreshed. The problem I encounter now is that when I customize a current limiting strategy and test it is feasible, I immediately modify the configuration file to delete the current custom current limiting configuration and push it to spring cloud config configuration center. The custom current limiting strategy that has been deleted is still available. Ah-even~）
by the way, my English is very very poor,Translator from Baidu,wish you understand.

**Comment by marcosbarbero** [comment]
I understand that you managed to make the rate limit work, and afterward, when you remove the configuration, the application somehow still keeps the configuration.

Did you try to change any other configuration on your application? I mean, something not related to the Rate Limiter. I need to know if any other configuration refreshes or keeps the same state.

**Comment by helloworld0929** [comment]
早上好哇，在我的配置中心上面除了与限流相关的配置没没有其他的配置，redis的配置使用的是默认的配置，如下：
(Good morning. There are no other configurations in my configuration center except the configurations related to current limitation. The configurations of redis use the default configurations as follows:)
```yml
zuul:
  routes:
    SDryRun:
      path: /internal/SDryRun
      url: forward:/internal
  ratelimit:
    repository: redis
    enabled: true
#    default-policy-list:
#      quota: 100
#      limit: 100
#      refresh-interval: 3
    policy-list:
      SDryRun:
        - limit: 5
          quota: 1
          refresh-interval: 60
          type:
            - url
```
（当我在linux上面修改application.yml后，通过Git推送到配置中心上面，配置中心是会自动更新推送的配置，这里上面您也说过啦是没有问题的，所以我仍然不知道删除的自定义限流策略仍然有效是什么原因造成的，就这些，谢谢。）
When I modify the application.yml on Linux and push it to the configuration center through Git, the configuration center will automatically update the push configuration. As you said above, there is no problem. So I still don't know what causes the deletion of the custom current limiting strategy to be effective. Thank you for that.

**Comment by marcosbarbero** [comment]
How is the config-server pushing the new config to the application? Do you have a webhook and event bus configured to the config-server and the application?

**Comment by helloworld0929** [comment]
Hi, by webhook.
It seems that I know what's wrong with me, but I'm not sure yet:

- In my application, I use @ConfigurationProperties(prefix = "zuul") and @Component annotation to load ratelimit properties from application.properties file. When I delete default-policy-list or policy-list from application.properties(this file in config-center), the properties is the same as before it was not deleted，

- But when I do add or modify operations, the configuration is synchronized~

Just wonder if is the problem of @ConfigurationProperties annnotation loading from the cache? 
If you know what the problem is, please let me know. Thank you very much.
+_+~~~~

**Comment by marcosbarbero** [comment]
I noticed that you are leaving the `policy-list` on your config, can you try removing it? Something like this;

```
zuul:
  ratelimit:
    enabled: true
    repository: redis
```

**Comment by helloworld0929** [comment]
Is a mistake, I use properties file,  just a sample above, In properties file, delete utils class delete by key, like the following code:

- Before delete:

```properties
zuul.ratelimit.policy-list.testService[0].refresh-interval=20
zuul.ratelimit.policy-list.testService[0].quota=20
```
- If delete `zuul.ratelimit.policy-list.testService[0].quota=20`, only the following code left.
```properties
zuul.ratelimit.policy-list.testService[0].refresh-interval=20
```
you can understand its a key-value. So that is not the bug is.

**Comment by marcosbarbero** [comment]
Can you show your full config for the `testService`? You need to have either a `quota` or a `limit` for a given configuration if you have neither of them your configuration should not work.

**Comment by helloworld0929** [comment]
Sorry, I cant give you the full config, I know the rules(I have try many ways to test your contribute, trust me I know that), just a sample, dont be serious, is a sample to show you is not the problems is.

**Comment by helloworld0929** [comment]
But I can show you a sample of full config here:

- Before delete:

```properties
eureka.client.service-url.defaultZone=http://******
eureka.instance.prefer-ip-address=true
spring.datasource.password=111
spring.datasource.url=jdbc:postgresql://*****:****/demodb?charSet=utf-8
spring.datasource.username=demodb
spring.jpa.database=postgresql
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
spring.jpa.generate-ddl=false
spring.jpa.hibernate.ddl-auto=create-drop
spring.jpa.properties.hibernate.cache.use_query_cache=false
spring.jpa.properties.hibernate.cache.use_second_level_cache=false
spring.jpa.properties.hibernate.generate_statistics=true
spring.jpa.show-sql=true
spring.profiles=staging
spring.rabbitmq.addresses=amqp://******
spring.redis.host=http://*********
spring.zipkin.base-url=http://*******
zuul.routes.testService.path=/internal/testService
zuul.routes.testService.url=forward:/
zuul.ratelimit.enabled=true
zuul.ratelimit.repository=jpa
zuul.ratelimit.default-policy-list[0].limit=2
zuul.ratelimit.default-policy-list[0].quota=2
zuul.ratelimit.default-policy-list[0].refresh-interval=6
zuul.ratelimit.default-policy-list[0].type[0].type=URL
zuul.ratelimit.policy-list.testService[0].limit=2
zuul.ratelimit.policy-list.testService[0].quota=2
zuul.ratelimit.policy-list.testService[0].refresh-interval=6
zuul.ratelimit.policy-list.testService[0].type[0].type=URL
```

- Then I delete policy-list:
```properties
eureka.client.service-url.defaultZone=http://******
eureka.instance.prefer-ip-address=true
spring.datasource.password=111
spring.datasource.url=jdbc:postgresql://*****:****/demodb?charSet=utf-8
spring.datasource.username=demodb
spring.jpa.database=postgresql
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
spring.jpa.generate-ddl=false
spring.jpa.hibernate.ddl-auto=create-drop
spring.jpa.properties.hibernate.cache.use_query_cache=false
spring.jpa.properties.hibernate.cache.use_second_level_cache=false
spring.jpa.properties.hibernate.generate_statistics=true
spring.jpa.show-sql=true
spring.profiles=staging
spring.rabbitmq.addresses=amqp://******
spring.redis.host=http://*********
spring.zipkin.base-url=http://*******
zuul.ratelimit.enabled=true
zuul.ratelimit.repository=jpa
zuul.ratelimit.default-policy-list[0].limit=2
zuul.ratelimit.default-policy-list[0].quota=2
zuul.ratelimit.default-policy-list[0].refresh-interval=6
zuul.ratelimit.default-policy-list[0].type[0].type=URL
```
That's all, thank you.

**Comment by helloworld0929** [comment]
I found that I was a verbose person and wanted to delete the question and repeat it.

**Comment by helloworld0929** [comment]
with more detail.

**Comment by marcosbarbero** [comment]
I meant only the `ratelimit` properties, thanks anyway :)
So, based on what you have shown I guess the problem isn't really a problem.

Your `service-specific-policy` configuration is the same as the `default-policy` configuration, what happens here is quite simple. Once you delete the `service-specific-policy` configuration the `default-policy` configuration takes over and as they are the same you see no difference in the result.

**Comment by marcosbarbero** [comment]
Closing due inactivity.
